src/persist/src/indexed/columnar.rs

```diff
@@ -0,0 +1,206 @@
+        ColumnarRecordsIter {
+            idx: 0,
+            records: self.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ColumnarRecordsIter<'a> {
+    idx: usize,
+    records: ColumnarRecordsRef<'a>,
+}
+
+impl<'a> Iterator for ColumnarRecordsIter<'a> {
+    type Item = ((&'a [u8], &'a [u8]), u64, isize);
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.records.len, Some(self.records.len))
+    }
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ret = self.records.get(self.idx);
+        self.idx += 1;
+        ret
+    }
+}
+
+impl<'a> ExactSizeIterator for ColumnarRecordsIter<'a> {}
+
+pub struct ColumnarRecordsBuilder {
+    records: ColumnarRecords,
+}
+
+impl Default for ColumnarRecordsBuilder {
+    fn default() -> Self {
+        Self {
+            records: ColumnarRecords {
+                len: 0,
+                key_offsets: Vec::new(),
+                key_data: Vec::new(),
+                val_offsets: Vec::new(),
+                val_data: Vec::new(),
+                timestamps: Vec::new(),
+                diffs: Vec::new(),
+            },
+        }
+    }
+}
+
+impl ColumnarRecordsBuilder {
+    pub fn len(&self) -> usize {
+        self.records.len
+    }
+
+    pub fn reserve(&mut self, additional: usize, key_size_guess: usize, val_size_guess: usize) {
+        self.records.key_offsets.reserve(additional);
+        self.records.key_data.reserve(additional * key_size_guess);
+        self.records.val_offsets.reserve(additional);
+        self.records.key_data.reserve(additional * val_size_guess);
+        self.records.timestamps.reserve(additional);
+        self.records.diffs.reserve(additional);
+    }
+
+    pub fn push(&mut self, record: ((&[u8], &[u8]), u64, isize)) {
+        let ((key, val), ts, diff) = record;
+        self.records.key_offsets.push(self.records.key_data.len());
+        self.records.key_data.extend_from_slice(key);
+        self.records.val_offsets.push(self.records.val_data.len());
+        self.records.val_data.extend_from_slice(val);
+        self.records.timestamps.push(ts);
+        self.records.diffs.push(diff);
+        self.records.len += 1;
+        debug_assert_eq!(self.records.key_offsets.len(), self.records.len);
+        debug_assert_eq!(self.records.val_offsets.len(), self.records.len);
+        debug_assert_eq!(self.records.timestamps.len(), self.records.len);
+        debug_assert_eq!(self.records.diffs.len(), self.records.len);
+    }
+
+    pub fn mut_diff(&mut self, idx: usize) -> Option<&mut isize> {
+        if idx >= self.records.len {
+            return None;
+        }
+        debug_assert_eq!(self.records.diffs.len(), self.records.len);
+        Some(&mut self.records.diffs[idx])
+    }
+
+    pub fn finish(mut self) -> ColumnarRecords {
+        self.records.key_offsets.push(self.records.key_data.len());
+        self.records.val_offsets.push(self.records.val_data.len());
```

Comment:
I think this is a good place to debug assert the ColumnarRecord validity postconditions